File: results/dnd-rest-benchmark/lifecycle-runs/20260731T063957Z_claude_claude-sonnet-5_javascript-node/routes/encounters.js
import { sendJson, parseJsonBody } from '../lib/http.js';
import { CR_XP, LEVEL_THRESHOLDS, countMultiplier, difficultyForXp } from '../lib/rules.js';

export function registerEncounterRoutes(router) {
  router.post('/v1/encounters/adjusted-xp', async (req, res) => {
    const body = await parseJsonBody(req, res);
    if (!body.ok) return;
    const party = body.data && body.data.party;
    const monsters = body.data && body.data.monsters;
    if (!Array.isArray(party) || !Array.isArray(monsters)) {
      sendJson(res, 400, { error: 'invalid request' });
      return;
    }

    let baseXp = 0;
    let monsterCount = 0;
    for (const m of monsters) {
      const xp = CR_XP[String(m.cr)];
      if (xp === undefined || typeof m.count !== 'number') {
        sendJson(res, 400, { error: 'unsupported challenge rating' });
        return;
      }
      baseXp += xp * m.count;
      monsterCount += m.count;
    }

    const multiplier = countMultiplier(monsterCount);
    const adjustedXp = baseXp * multiplier;

    const thresholds = { easy: 0, medium: 0, hard: 0, deadly: 0 };
    for (const member of party) {
      const t = LEVEL_THRESHOLDS[member.level];
      if (!t) {
        sendJson(res, 400, { error: 'unsupported party level' });
        return;
      }
      thresholds.easy += t.easy;
      thresholds.medium += t.medium;
      thresholds.hard += t.hard;
      thresholds.deadly += t.deadly;
    }

    sendJson(res, 200, {
      base_xp: baseXp,
      monster_count: monsterCount,
      multiplier,
      adjusted_xp: adjustedXp,
      difficulty: difficultyForXp(adjustedXp, thresholds),
      thresholds,
    });
  });

  router.post('/v1/initiative/order', async (req, res) => {
    const body = await parseJsonBody(req, res);
    if (!body.ok) return;
    const combatants = body.data && body.data.combatants;
    if (!Array.isArray(combatants)) {
      sendJson(res, 400, { error: 'invalid request' });
      return;
    }
    const scored = combatants.map((c) => ({
      name: c.name,
      dex: c.dex,
      score: c.roll + c.dex,
    }));
    scored.sort((a, b) => {
      if (b.score !== a.score) return b.score - a.score;
      if (b.dex !== a.dex) return b.dex - a.dex;
      return a.name < b.name ? -1 : a.name > b.name ? 1 : 0;
    });
    sendJson(res, 200, {
      order: scored.map((c) => ({ name: c.name, score: c.score })),
    });
  });
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T060905Z_claude_claude-sonnet-5_python-django/dndsite/views/encounters.py
"""Adjusted-XP difficulty calculation and initiative ordering. Stateless — no persistence."""

import json

from django.http import JsonResponse
from django.views.decorators.csrf import csrf_exempt

from dndsite.rules import CR_XP, encounter_difficulty, encounter_multiplier, numeric, party_xp_thresholds
from dndsite.views._util import json_body


@csrf_exempt
def adjusted_xp(request):
    if request.method != "POST":
        return JsonResponse({"error": "method not allowed"}, status=405)
    try:
        body = json_body(request)
        party = body["party"]
        monsters = body["monsters"]
    except (json.JSONDecodeError, KeyError, TypeError):
        return JsonResponse({"error": "invalid request"}, status=400)

    try:
        base_xp = 0
        monster_count = 0
        for monster in monsters:
            cr = str(monster["cr"])
            count = int(monster["count"])
            if cr not in CR_XP or count < 0:
                return JsonResponse({"error": "invalid monster"}, status=400)
            base_xp += CR_XP[cr] * count
            monster_count += count

        thresholds = party_xp_thresholds(party)
        if thresholds is None:
            return JsonResponse({"error": "unsupported level"}, status=400)
    except (KeyError, TypeError, ValueError):
        return JsonResponse({"error": "invalid request"}, status=400)

    multiplier = encounter_multiplier(monster_count)
    adjusted = numeric(base_xp * multiplier)
    difficulty = encounter_difficulty(adjusted, thresholds)

    return JsonResponse(
        {
            "base_xp": base_xp,
            "monster_count": monster_count,
            "multiplier": multiplier,
            "adjusted_xp": adjusted,
            "difficulty": difficulty,
            "thresholds": thresholds,
        }
    )


@csrf_exempt
def initiative_order(request):
    if request.method != "POST":
        return JsonResponse({"error": "method not allowed"}, status=405)
    try:
        body = json_body(request)
        combatants = body["combatants"]
    except (json.JSONDecodeError, KeyError, TypeError):
        return JsonResponse({"error": "invalid request"}, status=400)

    try:
        scored = []
        for combatant in combatants:
            name = combatant["name"]
            dex = combatant["dex"]
            roll = combatant["roll"]
            score = roll + dex
            scored.append({"name": name, "dex": dex, "score": score})
    except (KeyError, TypeError):
        return JsonResponse({"error": "invalid request"}, status=400)

    scored.sort(key=lambda c: (-c["score"], -c["dex"], c["name"]))

    order = [{"name": c["name"], "score": numeric(c["score"])} for c in scored]

    return JsonResponse({"order": order})
